doc/quaqc.1
.\" Automatically generated by Pandoc 3.3
.\"
.TH "QUAQC" "1" "" "Version 1.5" "quaqc User Manual"
.SH NAME
quaqc \- Quick ATAC\-seq QC
.SH SYNOPSIS
quaqc [options] file1.bam [file2.bam [\&...]]
.SH DESCRIPTION
quaqc allows for ATAC\-seq\-specific quality control and read filtering
of NGS data with minimal processing time and extremely low memory
overhead.
Any number of samples can be processed, using multiple threads if
desired.
quaqc outputs a comprehensive set of aligned read metrics, including
alignment size, fragment size, percent duplicates, mapq scores, read
depth, GC content, and others.
Although designed for ATAC\-seq data, quaqc can also be used for other
unspliced DNA\-seq experiments (such as ChIP\-seq) as many of the
metrics are related to genereal sequencing quality.
Non\-nuclear aligned reads can also be separately considered if so
desired, helpful when determining the level of chloroplast contamination
in plant ATAC\-seq.
.PP
When aligning ATAC\-seq reads, it is recommended not to perform any
filtering before using quaqc, as well as marking duplicates and fixing
mate coordinates for paired end data.
The BAM must also be coordinate sorted.
This can all be accomplished using \f[CR]samtools\f[R] in the following
manner during alignment:
.IP
.EX
[bowtie2/etc ... ] \[rs]
    | samtools fixmate \-m \-u \- \- \[rs]
    | samtools sort \-u \- \[rs]
    | samtools markdup \-\-write\-index \- file.bam
.EE
.SH OPTIONS
.TP
\f[B]\-m\f[R] \f[I]STR\f[R], \f[B]\-\-mitochondria\f[R]=\f[I]STR\f[R]
Comma\-separated mitochondrial reference names (no spaces).
This causes quaqc to skip reads mapping to these references for certain
metrics.
The remaining metrics will be merged for all matching references under a
single set of metrics for \[lq]mitochondrial\[rq] reads.
By default quaqc checks for references with names matching to chrM,
ChrM, Mt and mitochondria.
To skip this check entirely, simply provide an empty string
(e.g.\ \f[CR]\-m \[aq] \[aq]\f[R]).
.TP
\f[B]\-p\f[R] \f[I]STR\f[R], \f[B]\-\-plastids\f[R]=\f[I]STR\f[R]
This option behaves the exact same as the \f[CR]\-\-mitochondria\f[R]
option.
Metrics for all matching references will be merged under a single set of
metrics for \[lq]plastidic\[rq] reads.
By default quaqc checks for references with names matching to chrC,
ChrC, Pt and chloroplast.
To skip this check entirely, simply provide an empty string
(e.g.\ \f[CR]\-p \[aq] \[aq]\f[R]).
.TP
\f[B]\-P\f[R] \f[I]FILE\f[R], \f[B]\-\-peaks\f[R]=\f[I]FILE\f[R]
Filename of a BED file containing ranges that will be use to calculate
the fraction of reads in peaks (FRIP).
The impact on the runtime of quaqc when using this option is generally
negligible.
Can be gzipped.
.TP
\f[B]\-T\f[R] \f[I]FILE\f[R], \f[B]\-\-tss\f[R]=\f[I]FILE\f[R]
Filename of a BED file containing ranges that will be used to construct
a vector of read densities around the center of the ranges.
Afterwards a final TSS enrichment score (TES) is calculated by dividing
the average read depth around the middle 10% of the ranges by the
average read depth in the first 25% of the ranges.
This function is strand\-aware.
If ranges are missing strand information, they are assumed to be on the
forward/positive strand.
The impact on the runtime of quaqc when using this option is generally
negligible.
Can be gzipped.
.TP
\f[B]\-n\f[R] \f[I]STR\f[R], \f[B]\-\-target\-names\f[R]=\f[I]STR\f[R]
Comma\-separated reference names that quaqc will be restricted to (no
spaces).
This can include reference names also provided to
\f[CR]\-\-mitochondria\f[R] and \f[CR]\-\-plastids\f[R].
The effective genome size for metric calculations will also be adjusted.
Using this option will also proportionally scale the runtime down to the
fraction of the total genome being scanned, since quaqc will simply skip
over the entire section of the BAM not containing matching reads.
.TP
\f[B]\-t\f[R] \f[I]FILE\f[R], \f[B]\-\-target\-list\f[R]=\f[I]FILE\f[R]
Filename of a BED file containing ranges that quaqc will be restricted
to.
The effective genome size for metric calculations will also be adjusted.
Cannot be used simultaneously with \f[CR]\-\-blacklist\f[R].
Can be gzipped.
To be used for QC, reads must be fully contained within these ranges.
.TP
\f[B]\-b\f[R] \f[I]FILE\f[R], \f[B]\-\-blacklist\f[R]=\f[I]FILE\f[R]
Filename of a BED file containing ranges in which all read aligning to
these regions are to be ignored by quaqc.
The effective genome size for metric calculations will also be adjusted.
Cannot be used simultaneously with \f[CR]\-\-target\-list\f[R].
Can be gzipped.
To be used for QC, reads must be fully outside these ranges.
.TP
\f[B]\-r\f[R] \f[I]STR\f[R], \f[B]\-\-rg\-names\f[R]=\f[I]STR\f[R]
Comma\-separated read group identifiers.
This will cause quaqc to discard reads which do not have matching read
group tags.
Cannot be used simultaneously with \f[CR]\-\-rg\-list\f[R].
If a BAM header does not contain any of the target read groups, the run
will end with an error.
.TP
\f[B]\-R\f[R] \f[I]STR\f[R], \f[B]\-\-rg\-list\f[R]=\f[I]STR\f[R]
Filename of a file containing read group identifiers, one per line.
This will cause quaqc to discard reads which do not have a matching read
group tag.
Cannot be used simultaneously with \f[CR]\-\-rg\-names\f[R].
If a BAM header does not contain any of the target read groups, the run
will end with an error.
Can be gzipped.
.TP
\f[B]\-\-rg\-tag\f[R]=\f[I]STR\f[R]
Filter reads using a different SAM tag than the read group (RG) tag.
This can be used, for example, to filter single\-cell BAMs using the
barcode (CB) tag.
Please note that the SAM spec requires all possible RG groups to be
encoded in the header, meaning quaqc can check that the provided groups
from \f[CR]\-\-rg\-list\f[R] exist in the BAM before starting.
No such check is performed when using a different tag, meaning mistakes
can lead to quaqc filtering out all reads.
.TP
\f[B]\-2\f[R], \f[B]\-\-use\-secondary\f[R]
Do not filter out alignments marked with the secondary alignment flag
for use in the calculation of metrics done after read filtering.
Useful when you wish to include multi\-mapped reads (such as when
\f[CR]bowtie2\f[R] is run with the \f[CR]\-k\f[R] flag).
.TP
\f[B]\-N\f[R], \f[B]\-\-use\-nomate\f[R]
Do not filter out paired\-end reads when it is properly mapped but its
mate read is not (for use in the calculation of metrics done after read
filtering).
Note that quaqc gets fragment and mate information from the individual
read metadata and SAM flags (i.e.\ it does not wait to see the
corresponding read to know if it will be properly mapped or calculate
the fragment size).
Depending on the sequence aligner which created the BAM, this
information may be absent without using \f[CR]samtools fixmate\f[R] or
equivalent.
If this is the case, using this argument may be required to prevent
quaqc from classifying all paired\-ends as filtered.
.TP
\f[B]\-d\f[R], \f[B]\-\-use\-dups\f[R]
Do not filter out duplicate reads for use in the calculation of metrics
done after read filtering.
Note that quaqc can only know if reads are duplicates based on their SAM
flags, which may require use of \f[CR]samtools markdup\f[R] or
equivalent.
.TP
\f[B]\-\-use\-chimeric\f[R]
Do not filter out alignments marked with the supplementary alignment
flag for use in the calculation of metrics done after read filtering.
Generally this argument should never be set outside of special cases.
.TP
\f[B]\-D\f[R], \f[B]\-\-use\-dovetails\f[R]
Do not filter out dovetailing alignments from paired reads.
Dovetailing is when a read end coordinate extends past the start
coordinate of its mate.
.TP
\f[B]\-\-no\-se\f[R]
Filter out single end (SE) reads, if a BAM happens to contain both SE
and PE reads.
.TP
\f[B]\-q\f[R] \f[I]INT\f[R], \f[B]\-\-mapq\f[R]=\f[I]INT\f[R]
Filter out reads with a MAPQ below this value for use in the calculation
of metrics done after read filtering.
The default value is 30.
Note that reads with MAPQ scores of 255 are classified as unscored and
will ignore this setting.
.TP
\f[B]\-\-min\-qlen\f[R]=\f[I]INT\f[R]
Filter out reads which cover fewer bases on the reference than this
value for use in the calculation of metrics done after read filtering.
The default value is 1.
.TP
\f[B]\-\-min\-flen\f[R]=\f[I]INT\f[R]
Filter out paired\-end reads with a fragment size below this value for
use in the calculation of metrics done after read filtering.
The default value is 1.
Note that this argument is effectively ignored when
\f[CR]\-\-use\-nomate\f[R] is set.
.TP
\f[B]\-\-max\-qlen\f[R]=\f[I]INT\f[R]
Filter out reads which cover more bases on the reference than this value
for use in the calculation of metrics done after read filtering.
The default value is 250.
.TP
\f[B]\-\-max\-flen\f[R]=\f[I]INT\f[R]
Filter out paired\-end reads with a fragment size higher than this value
for use in the calculation of metrics done after read filtering.
The default value is 2000.
Note that this argument is effectively ignored when
\f[CR]\-\-use\-nomate\f[R] is set.
.TP
\f[B]\-A\f[R], \f[B]\-\-use\-all\f[R]
Do not apply ANY filters to mapped reads.
Note that unmapped reads are always ignored, no matter which options are
set.
.TP
\f[B]\-\-max\-depth\f[R]=\f[I]INT\f[R]
The maximum allowed value for the read depth histogram.
Depths with a higher value than this will be counted as this value.
The default is 100,000.
Be aware that this value is used directly during memory allocation.
To estimate how much memory will be used for the depths histogram,
multiply the value by 8 bytes (e.g.\ the default uses 800 kB).
If quaqc detects depth values equal to the set maximum during QC report
generation, it will add a note warning that some depth values may have
been truncated (which will result in incorrect stats).
The average depth can be calculated without this data so is unaffected
by this issue.
.TP
\f[B]\-\-max\-qhist\f[R]=\f[I]INT\f[R]
The maximum allowed value for the covered bases per read histogram.
Density values greater than this value will be reported as this value.
The default is the value of \f[CR]\-\-max\-qlen\f[R].
Be aware that this value is used directly during memory allocation.
To estimate how much memory will be used for the read size histogram,
multiply the value by 4 bytes (e.g.\ the default uses 1 kB).
If quaqc detects size values equal to the set maximum during QC report
generation, it will add a note warning that some size values may have
been truncated (which will result in incorrect stats).
The average size can be calculated without this data so is unaffected by
this issue.
.TP
\f[B]\-\-max\-fhist\f[R]=\f[I]INT\f[R]
The maximum allowed value for the fragment size histogram.
Density values greater than this value will be reported as this value.
The default is the value of \f[CR]\-\-max\-flen\f[R].
Be aware that this value is used directly during memory allocation.
To estimate how much memory will be used for the fragment size
histogram, multiply the value by 4 bytes (e.g.\ the default uses 8 kB).
If quaqc detects size values equal to the set maximum during QC report
generation, it will add a note warning that some size values may have
been truncated (which will result in incorrect stats).
The average size can be calculated without this data so is unaffected by
this issue.
.TP
\f[B]\-\-tss\-size\f[R]=\f[I]INT\f[R]
The size of the density vector range generated when \f[CR]\-\-tss\f[R]
is set, in bases.
Ranges are first centered at their midpoints, then resized in both
directions to a final width of the set value.
The default is 2000.
Be aware that this value is used directly during memory allocation.
To estimate how much memory will be used for the TSS density values,
multiply the value by 4 bytes (e.g.\ the default uses 8 kB).
.TP
\f[B]\-\-tss\-qlen\f[R]=\f[I]INT\f[R]
The final size of adjusted read coordinates when generating the read
density values when \f[CR]\-\-tss\f[R] is set.
Reads are first set to size 1 (anchored from their five\-prime ends),
then resized in both directions to a final width of the set value.
The default is 100.
To prevent read resizing and instead use the actual coordinates of the
reads, set this value to 0.
.TP
\f[B]\-\-tss\-tn5\f[R]
When resizing the reads as described in the \f[CR]\-\-tss\-qlen\f[R]
option, adjust the read five\-prime coordinates forward 4 bases (to
center the coordinate in the middle of the Tn5 transposase binding
site).
This option is ignored when \f[CR]\-\-tss\-qlen\f[R] is set to 0.
.TP
\f[B]\-\-omit\-gc\f[R]
Skip GC content metrics.
This can shave off a small percentage of the runtime for regular short
read experiments (<10%).
The savings may be more substantial for long read experiments, as quaqc
has to iterate over every base in the alignments to count GC bases.
.TP
\f[B]\-\-omit\-depth\f[R]
Skip generation of the read depths histogram.
This can shave off a small percentage of the runtime for regular short
read experiments (<10%).
The savings may be more substantial for long read experiments, as quaqc
has to iterate over the entire alignment length to count per\-base
depths.
.TP
\f[B]\-f\f[R], \f[B]\-\-fast\f[R]
Set \f[CR]\-\-omit\-gc\f[R] and \f[CR]\-\-omit\-depth\f[R], thus
skipping the two metric which require iterating over the entire read
lengths.
Together this can shave off about 15% of the runtime for regular short
read experiments.
The savings may be more substantial for long read experiments.
.TP
\f[B]\-\-lenient\f[R]
Set \f[CR]\-\-use\-nomate\f[R], \f[CR]\-\-use\-dups\f[R],
\f[CR]\-\-use\-dovetails\f[R], and \f[CR]\-\-mapq=10\f[R].
This relaxes the filtering parameters, allowing a greater number of
reads to be counted for QC.
.TP
\f[B]\-\-strict\f[R]
Set \f[CR]\-\-min\-flen=50\f[R], \f[CR]\-\-max\-flen=150\f[R], and
\f[CR]\-\-mapq=40\f[R].
This restricts the filtering parameters to keep only the highest quality
reads.
.TP
\f[B]\-\-nfr\f[R]
Set \f[CR]\-\-no\-se\f[R], \f[CR]\-\-max\-flen=120\f[R], and
\f[CR]\-\-tss\-tn5\f[R].
These filters enrich for reads found within nucleosome free regions
(NFR), as well as shifting the start sites to account for the Tn5
transposase insertion.
.TP
\f[B]\-\-nbr\f[R]
Set \f[CR]\-\-no\-se\f[R], \f[CR]\-\-min\-flen=150\f[R],
\f[CR]\-\-max\-flen=1000\f[R], and \f[CR]\-\-tss\-qlen=0\f[R].
These filters enrich for reads in nucleosome bound regions (NBR).
In addition, the read coordinates are maintained as is when generating
the TSS pileup.
.TP
\f[B]\-\-footprint\f[R]
Set \f[CR]\-\-tss\-qlen=1\f[R], \f[CR]\-\-tss\-size=501\f[R], and
\f[CR]\-\-tss\-tn5\f[R].
This generates a smaller TSS pileup with single base pair resolution of
Tn5 transposase insertion frequency.
.TP
\f[B]\-\-chip\f[R]
Set \f[CR]\-\-tss\-qlen=0\f[R] and \f[CR]\-\-tss\-size=5001\f[R].
Additionally, any BED file provided with the \f[CR]\-\-peaks\f[R] option
is used for generating the pileup (which is normally generated from
\f[CR]\-\-tss\f[R]).
.TP
\f[B]\-o\f[R] \f[I]DIR\f[R], \f[B]\-\-output\-dir\f[R]=\f[I]DIR\f[R]
Directory where the QC reports will be saved.
By default, these are saved in the same directory as the input BAMs.
.TP
\f[B]\-O\f[R] \f[I]STR\f[R], \f[B]\-\-output\-ext\f[R]=\f[I]STR\f[R]
Filename extension of the QC report, replacing the previous
\[lq].bam\[rq] of the input BAMs.
By default \[lq].quaqc.txt\[rq] is used.
.TP
\f[B]\-0\f[R], \f[B]\-\-no\-output\f[R]
Suppress the generation of QC reports.
.TP
\f[B]\-J\f[R] \f[I]FILE\f[R], \f[B]\-\-json\f[R]=\f[I]FILE\f[R]
Save all QC reports for all samples into a single JSON file for further
processing.
This format, while not intended to be human readable, contains
additional data such as the full alignment size, fragment size, GC
content, mapq, and read depth histograms, as well as the TSS pileup.
To save to standard output, provide \f[CR]\-J\-\f[R].
To compress the output JSON, add the \[lq].gz\[rq] extension to the
supplied filename.
.TP
\f[B]\-S\f[R], \f[B]\-\-keep\f[R]
Save the nuclear reads passing all filters in a new BAM.
This will significantly slow down quaqc.
.TP
\f[B]\-k\f[R] \f[I]DIR\f[R], \f[B]\-\-keep\-dir\f[R]=\f[I]DIR\f[R]
By default, when \f[CR]\-\-keep\f[R] is set a new filtered BAM is
created in the same directory as the input BAM.
Setting this will change the final directory where the new BAM will be
written.
.TP
\f[B]\-K\f[R] \f[I]STR\f[R], \f[B]\-\-keep\-ext\f[R]=\f[I]STR\f[R]
By default, when \f[CR]\-\-keep\f[R] is set a new filtered BAM is
created with the text \[lq].filt.bam\[rq] appended to the file name.
Use this argument to change it.
If an existing \[lq].bam\[rq] or \[lq].cram\[rq] extension exists, it
will be stripped.
.TP
\f[B]\-B\f[R], \f[B]\-\-bedGraph\f[R]
Output a Gzipped bedGraph of the alignments passing all filters within
target regions.
Only limited memory is used to store bedGraph records which are output
on the fly.
This means that some consecutive positions with identical scores will
sometimes be present as distinct ranges if quaqc ran out of memory to
store such records simultaneously.
.TP
\f[B]\-\-bedGraph\-qlen\f[R]=\f[I]INT\f[R]
When outputting the alignments in bedGraph format, they are resized from
the 5\-prime position.
This option controls the final size of the 5\-prime centered alignment.
To instead use the original start and end coordinates of the alignment,
set this option to 0.
.TP
\f[B]\-\-bedGraph\-tn5\f[R]
Shift the 5\-prime alignment coordinates of each read when generating
the bedGraph to account for the transposase offset (+4/\-5), as per the
\f[CR]\-\-tss\-tn5\f[R] option.
.TP
\f[B]\-\-bedGraph\-dir\f[R]=\f[I]DIR\f[R]
As per the \f[CR]\-\-keep\-dir\f[R] option, change the output directory
of the bedGraph files.
.TP
\f[B]\-\-bedGraph\-ext\f[R]=\f[I]STR\f[R]
As per the \f[CR]\-\-keep\-ext\f[R] option, change the default bedGraph
filename extension.
The bedGraph will always be Gzipped, so not including \[lq].gz\[rq] will
still lead to a compressed file.
.TP
\f[B]\-\-bed\f[R]
Output a Gzipped BED6 file of the alignments passing all filters within
target regions.
The name field contains the read name, with an additional \[lq]/1\[rq]
or \[lq]/2\[rq] appended to the first and second mates of PE reads.
Alignments are printed in the same order they are present in the BAM,
thus they can be assumed to be coordinate sorted.
.TP
\f[B]\-\-bed\-ins\f[R]
Modify the BED output to instead print out the exact insertion
coordinates (i.e.\ the 5\-prime ends of the alignments) in BED3 format.
In order to maintain a coordinate sorting, the order of insertion
coordinates will not match the original order of alignments in the BAM.
.TP
\f[B]\-\-bed\-tn5\f[R]
Shift the 5\-prime alignment coordinates of each read when generating
then BED file to account for the transposase offset (+4/\-5), as per the
\f[CR]\-\-tss\-tn5\f[R] option.
.TP
\f[B]\-\-bed\-dir\f[R]
As per the \f[CR]\-\-keep\-dir\f[R] option, change the output directory
of the BED files.
.TP
\f[B]\-\-bed\-ext\f[R]=\f[I]STR\f[R]
As per the \f[CR]\-\-keep\-ext\f[R] option, change the default BED
filename extension.
The BED files will always be Gzipped, so not including \[lq].gz\[rq]
will still lead to a compressed file.
.TP
\f[B]\-\-quant\f[R]=\f[I]FILE\f[R]
Output a TSV file of counts of reads in the provided peaks
(\f[CR]\-\-peaks\f[R]) BED file.
Ending the filename with \[lq].gz\[rq] will cause quaqc to gzip the
file.
The first column in the file are peak names.
If these are lacking from the peaks BED file, or if the name field has
the empty placeholder \[lq].\[rq], then the peak coordinates in the
format \[lq]chromosome:start\-end\[rq] are used instead.
After this, the following columns contain the counts for the samples, in
the order that they are provided to quaqc.
Each count comes from a passing nuclear read which at least partially
overlaps a range in the peaks BED.
.TP
\f[B]\-\-quant\-ins\f[R]
Instead of simply counting a read if it partially overlaps anywhere in
its range, a read only counts as overlapping if its 5\-prime coordinates
(the insertion sites) are within a peak range.
.TP
\f[B]\-\-quant\-tn5\f[R]
Shift the 5\-prime alignment coordinates of each read to account for the
transposase offset (+4/\-5), as per the \f[CR]\-\-tss\-tn5\f[R] option,
before it is used to check for an overlap with a peak range.
This option is applied before \f[CR]\-\-quant\-ins\f[R], if that is
used.
.TP
\f[B]\-\-quant\-pn\f[R]
By default, the full filename and any directories included in the
provided sample paths are printed in the output counts TSV.
When this option is set, any directory names in the provided path are
removed and if the file ends with \[lq].bam\[rq] the extension is
stripped, providing a prettier sample name in the output file.
.TP
\f[B]\-\-tn5\-fwd\f[R]=\f[I]INT\f[R]
Alter the default value used to shift the 5\-prime ends of reads when
either \f[CR]\-\-tss\-tn5\f[R], \f[CR]\-\-bedGraph\-tn5\f[R], or
\f[CR]\-\-bed\-tn5\f[R] are used.
This value is added to the start coordinate of forward strand reads.
.TP
\f[B]\-\-tn5\-rev\f[R]=\f[I]INT\f[R]
Alter the default value used to shift the 5\-prime ends of reads when
either \f[CR]\-\-tss\-tn5\f[R], \f[CR]\-\-bedGraph\-tn5\f[R], or
\f[CR]\-\-bed\-tn5\f[R] are used.
This value is substracted from the end coordinate of reverse strand
reads.
.TP
\f[B]\-j\f[R] \f[I]INT\f[R], \f[B]\-\-threads\f[R]=\f[I]INT\f[R]
Set the number of child threads used to process input BAMs.
At minimum, one child thread is launched (meaning quaqc technically uses
two threads, though not simultaneously), and at maximum, one child
thread per sample is launched (in addition to the main parent thread).
All of the data structures are duplicated for each new thread, meaning
memory usage will increase linearly with increasing thread count.
When using default settings, the \f[CR]\-\-max\-depth\f[R] option has
the biggest impact on memory growth.
Set this to a lower value to mitigate this.
.TP
\f[B]\-i\f[R] \f[I]STR\f[R], \f[B]\-\-title\f[R]=\f[I]STR\f[R]
Assign a title to the run.
All output reports will contain this title.
.TP
\f[B]\-c\f[R], \f[B]\-\-continue\f[R]
If set when processing more than one input file, quaqc will keep running
if it encounters errors processing individual files (e.g.\ one file is
unsorted).
.TP
\f[B]\-v\f[R], \f[B]\-\-verbose\f[R]
Print progress messages during runtime.
This flag can be used a second time to further increase verbosity.
.TP
\f[B]\-\-version\f[R]
Print the version number of quaqc to \f[CR]stdout\f[R] and exit.
.TP
\f[B]\-h\f[R], \f[B]\-\-help\f[R]
Print a help message with a brief description of all available commands.
.SH BUGS
Please report bugs on GitHub: \c
.UR https://github.com/bjmt/quaqc/issues
.UE \c
.SH AUTHOR
quaqc was created by Benjamin Jean\-Marie Tremblay.


NEWS
CHANGES IN VERSION 1.5 (2025-09-12)
-----------------------------------

* Added read quantification (--quant). This outputs a TSV of read
  counts, with peaks as rows and samples as columns. The peaks can
  be provided via --peaks. Quantification can be at the insertion
  level via --quant-ins.

* Fixed a couple of bugs with --bed options.

CHANGES IN VERSION 1.4 (2025-08-29)
-----------------------------------

* Added BED output (--bed). The behaviour of this output is similar
  to the default behaviour of `bedtools bamtobed`. Additionally, the
  BED can contain the Tn5 insertion coordinates instead of the
  original alignment coordinates by adding the --bed-ins option.

* Fixed typos & formatting errors in output reports

* More --bedGraph bugfixes

* Fix compilation of included zlib library with newer LLVM versions

CHANGES IN VERSION 1.3 (2025-05-04)
-----------------------------------

* Added the ability to filter reads using alternate SAM tags instead of
  read groups (RG) via --rg-tag. 

* Fixed some bugs in the --bedGraph implementation.

CHANGES IN VERSION 1.2 (2025-05-03)
-----------------------------------

* Added bedGraph output (--bedGraph). The behaviour of the output pileup
  is similar to the pileup created by the --tss option, with the ability
  to either make a pileup from the exact read coordinates or centered
  around resized 5-prime coordinates.

CHANGES IN VERSION 1.1 (2024-06-13)
-----------------------------------

* Added --strict flag: Use strict filtering parameters from manuscript

* Overlapping ranges provided to --tss are now removed to prevent imbalanced
  pileups due to reads only being counted for the first overlapping range
  even if multiple overlapping ranges exist

* Fixed a bug with read overlap detection for --tss option



quaqc.R
#!/usr/bin/env Rscript
#
#   quaqc: QUick Atac-seq Quality Control
#   Copyright (C) 2024  Benjamin Jean-Marie Tremblay
#
#   This program is free software: you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation, either version 3 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program.  If not, see <https://www.gnu.org/licenses/>.
#

quaqc.R_version <- "0.1"
quaqc.R_year <- "2024"

warning("Note that quaqc.R is not yet complete! Use at your own risk.")

#-------------------------------------------------------------------------------
# Choose whether to launch app in interactive sessions.

# launchApp <- !interactive()
launchApp <- TRUE

#-------------------------------------------------------------------------------
# Load packages

if (launchApp) message("Loading dependencies ...")

pkgs <- c("shiny"
  , "shinydashboard"
  , "ggplot2"
  , "jsonlite"
  , "DT"
)

checkPkg <- function(x) {
  if (!requireNamespace(x, quietly = TRUE)) x else c()
}

missing.packages <- c()
for (pkg in pkgs) missing.packages <- c(missing.packages, checkPkg(pkg))

if (length(missing.packages)) {
  message("Missing ", length(missing.packages), " packages.")
  answer <- menu(c("Yes", "No"), title = "Do you wish to install missing packages?")
  if (answer == 1) {
    message("Installing packages.")
    install.packages(missing.packages)
  } else {
    message("Please install the following packages:")
    message("    c(", paste0(paste0("\"", missing.packages, "\""), collapse = ", "), ")")
    quit(save = "no")
  }
}

for (pkg in pkgs) suppressPackageStartupMessages(library(pkg, character.only = TRUE))

#-------------------------------------------------------------------------------
# Functions

readQuaqcReport <- function(x) {
  f <- gzfile(x, "rt")
  on.exit(close(f))
  fromJSON(readLines(f), simplifyDataFrame = FALSE)
}

multiReadQuaqcReports <- function(x) {
  lapply(x, function(y) {
    tryCatch(readQuaqcReport(y), error = function(e) NULL)
  })
}

checkReport <- function(x) {
  if (any(!c("quaqc_version", "quaqc_reports", "quaqc_time_end") %in% names(x))) {
    FALSE
  } else {
    TRUE
  }
}

reportToDf <- function(f, x) {
  if (is.null(x) || !checkReport(x)) {
    data.frame(row.names = NULL, check.names = FALSE,
      File = f,
      Title = "",
      Status = "BAD INPUT",
      Version = NA,
      Samples = NA,
      Date = NA
    )
  } else {
    data.frame(row.names = NULL, check.names = FALSE,
      File = f,
      Title = x$quaqc_run_title,
      Status = "OK",
      Version = x$quaqc_version,
      Samples = length(x$quaqc_reports),
      Date = x$quaqc_time_end
    )
  }
}

reportToParams <- function(x, i) {
  if (is.null(i)) {
    data.frame(row.names = NULL, check.names = FALSE,
      Flag = c("--mitochondria", "--plastids", "--peaks", "--tss",
        "--target-names", "--target-list", "--blacklist", "--rg-names",
        "--rg-list", "--use-secondary", "--use-nomate", "--use-dups",
        "--use-chimeric", "--use-dovetails", "--no-se", "--mapq",
        "--min-qlen", "--min-flen", "--max-qlen", "--max-flen", "--use-all",
        "--max-depth", "--max-qhist", "--max-fhist", "--tss-size", 
        "--tss-qlen", "--tss-tn5", "--omit-gc", "--omit-depth",
        "--fast", "--lenient", "--nfr", "--nbr", "--footprint", "--chip",
        "--output-dir", "--output-ext", "--no-output", "--json",
        "--keep", "--keep-dir", "--keep-ext", "--threads", "--title",
        "--continue", "--verbose"),
      Value = NA
    )
  } else {
    r <- x[[i]]$quaqc_params
    data.frame(row.names = NULL, check.names = FALSE,
      Flag = c("--mitochondria", "--plastids", "--peaks", "--tss",
        "--target-names", "--target-list", "--blacklist", "--rg-names",
        "--rg-list", "--use-secondary", "--use-nomate", "--use-dups",
        "--use-chimeric", "--use-dovetails", "--no-se", "--mapq",
        "--min-qlen", "--min-flen", "--max-qlen", "--max-flen", "--use-all",
        "--max-depth", "--max-qhist", "--max-fhist", "--tss-size", 
        "--tss-qlen", "--tss-tn5", "--omit-gc", "--omit-depth",
        "--fast", "--lenient", "--nfr", "--nbr", "--footprint", "--chip",
        "--output-dir", "--output-ext", "--no-output", "--json",
        "--keep", "--keep-dir", "--keep-ext", "--threads", "--title",
        "--continue", "--verbose"),
      Value = NA
    )
  }
}

#-------------------------------------------------------------------------------
# App UI

ui <- dashboardPage(
  dashboardHeader(title = "quaqc"),
  dashboardSidebar(
    sidebarMenu(
      menuItem("Load results", tabName = "load"),
      menuItem("Explore results", tabName = "results"),
      menuItem("About", tabName = "about")
    )
  ),
  dashboardBody(
    tabItems(
      tabItem(tabName = "load",
        box(width = NULL, status = "primary", solidHeader = TRUE,
          title = "Load quaqc JSON reports (.json, .json.gz)",
          fileInput("report_files", NULL, multiple = TRUE,
            accept = c(".json", ".json.gz")
          ),
          dataTableOutput("report_files_table")
        ),
        box(width = NULL, status = "primary", solidHeader = TRUE,
          title = "Parameters of loaded JSON report",
          dataTableOutput("report_parameters")
        )
      ),
      tabItem(tabName = "results"),
      tabItem(tabName = "about")
    )
  )
)

#-------------------------------------------------------------------------------
# App server

server <- function(input, output, session) {

  reports <- reactiveValues(
    files = NULL, data = NULL, selected = NULL
  )

  output$report_files_table <- renderDataTable(datatable({
    req(input$report_files)
    reports$files <- input$report_files
    reports$data <- multiReadQuaqcReports(input$report_files$datapath)
    reportRows <- mapply(function(x, y) reportToDf(x, y),
      reports$files$name, reports$data, SIMPLIFY = FALSE)
    reportTable <- do.call(rbind, reportRows)
    reportTable[order(reportTable$Status, decreasing = TRUE), ]
  }, selection = list(mode = "single", selected = 1), rownames = FALSE,
    options = list(dom = "t", columnDefs = list(list(className = "dt-center", targets = "_all"))))
  )

  output$report_parameters <- renderDataTable(datatable({
    # req(input$report_files)
    reportToParams(reports$data, input$report_files_table_rows_selected)
  }, rownames = FALSE, selection = "none",
    options = list(ordering = FALSE, dom = "t", pageLength = 999,
      columnDefs = list(list(className = "dt-left", targets = "_all"))))
  )

}

#-------------------------------------------------------------------------------
# Launch app

if (launchApp) {
  message("All OK. Launching App.")
  runApp(list(ui = ui, server = server), launch.browser = TRUE, quiet = FALSE)
}

